File: 18-Binary-Tree-Part-II/python/04_balanced_binary_tree.py
from typing import Optional

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return True

        heights = {}
        stack = [root]

        while stack:
            node = stack.pop()

            left_done = node.left is None or node.left in heights
            right_done = node.right is None or node.right in heights

            if not left_done or not right_done:
                stack.append(node)
                if not left_done:
                    stack.append(node.left)
                if not right_done:
                    stack.append(node.right)
                continue

            left_height = 0 if node.left is None else heights[node.left]
            right_height = 0 if node.right is None else heights[node.right]

            if abs(left_height - right_height) > 1:
                return False

            heights[node] = 1 + max(left_height, right_height)

        return True

File: 18-Binary-Tree-Part-II/cpp/04-balanced-binary-tree.cpp
#include <unordered_map>
#include <stack>
#include <cmath>
#include <algorithm>

using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isBalanced(TreeNode* root) {
        if (!root) return true;

        unordered_map<TreeNode*, int> heights;
        stack<TreeNode*> st;
        st.push(root);

        while (!st.empty()) {
            TreeNode* node = st.top();
            st.pop();

            bool leftDone = (node->left == nullptr) || heights.count(node->left);
            bool rightDone = (node->right == nullptr) || heights.count(node->right);

            if (!leftDone || !rightDone) {
                st.push(node);
                if (!leftDone) st.push(node->left);
                if (!rightDone) st.push(node->right);
                continue;
            }

            int leftHeight = (node->left == nullptr) ? 0 : heights[node->left];
            int rightHeight = (node->right == nullptr) ? 0 : heights[node->right];

            if (abs(leftHeight - rightHeight) > 1) {
                return false;
            }

            heights[node] = 1 + max(leftHeight, rightHeight);
        }

        return true;
    }
};
